added types for Promise, Map, fetch API, Map and history

aholtkamp opened the pull request.

aholtkamp committed:
added types for Promise, Map, fetch API, Map and history

fwienber commented on jangaroo-browser/src/main/joo/js/AbortSignal.as:
```diff
@@ -0,0 +1,27 @@
+
+/**
+ * A signal object that allows you to communicate with a DOM request (such as a Fetch) and abort it if required via an
+ * {@link AbortController} object.
```
Unfortunately, there is no @link in ActionScript.
Our convention is to use something like a footnote style:

    ... →<code>link-ref|link-text</code> ...

    @see link-ref [link-text]

fwienber replied:
I wouldn't use this technique to link String or Array. Too many @see to such well-known classes don't really improve readability.

fwienber commented on jangaroo-browser/src/main/joo/js/Promise.as:
```diff
@@ -0,0 +1,61 @@
+package js {
+
+/**
+ * Represents the completion of an asynchronous operation
```
```suggestion
 * Represents the completion of an asynchronous operation.
```

fwienber commented on jangaroo-browser/src/main/joo/js/Body.as:
```diff
@@ -0,0 +1,48 @@
+package js {
+
+/**
+ *
```
```suggestion
 * The <code>Body</code> mixin of the →<code>fetch</code> API represents the body of the response/request,
 * allowing you to declare what its content type is and how it should be handled.
 *
 * <p><code>Body</code> is implemented by both <code>Request</code> and <code>Response</code>.
 * This provides these objects with an associated body (a stream), a used flag (initially unset),
 * and a MIME type (initially the empty byte sequence).
 * </p>
 *
 * @see js.#fetch() fetch()
```

fwienber commented on jangaroo-browser/src/main/joo/js/Request.as:
```diff
@@ -0,0 +1,103 @@
+   * Creates a new Request object.
+   *
+   * @param input Defines the resource that you wish to fetch. This can either be:
+   *              - A {@link String} containing the direct URL of the resource you want to fetch.
+   *              - A {@link Request} object, effectively creating a copy. Note the following behavioural updates to
+   *                retain security while making the constructor less likely to throw exceptions:
+   *                - If this object exists on another origin to the constructor call, the Request.referrer is stripped
+   *                  out.
+   *                - If this object has a Request.mode of navigate, the mode value is converted to same-origin.
```
```suggestion
   *            <ul>
   *              <li>A {@link String} containing the direct URL of the resource you want to fetch.</li>
   *              <li>A {@link Request} object, effectively creating a copy. Note the following behavioural updates to
   *                retain security while making the constructor less likely to throw exceptions:
   *                <ul>
   *                  <li>If this object exists on another origin to the constructor call, the Request.referrer is
   *                  stripped out.</li>
   *                  <li>If this object has a Request.mode of navigate, the mode value is converted to
   *                  same-origin.</li>
   *                </ul>
   *              </li>
   *            </ul>
```

fwienber commented on jangaroo-browser/src/main/joo/js/FormData.as:
```diff
@@ -0,0 +1,86 @@
+
+  /**
+   * Deletes a key/value pair from a FormData object.
+   *
```
```suggestion
   * Aliased by Jangaroo with an underscore suffix to avoid clash with ActionScript keyword
   * <code>delete</code>. To make this alias work, it is added to the class prototype.
   *
```

Same for all other ...`_` aliases, please.

fwienber reviewed (changes requested):
Cool! Looks like a lot of work.
The requested changes are hopefully done quickly.

aholtkamp committed:
removed links for String and Array in as-doc

aholtkamp committed:
@link => <code> and @see

on the fly fixed too long lines

aholtkamp committed:
added missing class asdoc and fixed typo

aholtkamp committed:
added hints for methods that were renamed due to clash with AS

aholtkamp committed:
converted bullet points into <ul><li>

aholtkamp committed:
fixed see

- changed to absolute class names
- added () to functions

fwienber committed:
Repair jangaroo-libs-doc ASDoc Build Process

It now generates ASDoc of all relevant libraries
when calling 'mvn site'.
There are still some ASDoc problems (at least on
this branch) that are fixed in upcoming commits.

fwienber committed:
Fix nested <code> elements in ASDoc of NaN.as

fwienber committed:
Fix multi-line code examples in TestCase.as: <code> -> <pre>

fwienber committed:
Fix ASDoc: encode "<" to &lt;

fwienber committed:
Workaround for @see reference to package-scope function

IDEA wants the simple syntax "js.fetch()", but that does not
work in "real" ASDoc.
ASDoc wants "js.#fetch()" to render the correct link, but then
renders the link test incorrectly as "js..fetch()", so we repeat
"js.fetch()" as link text.
Now we end up with one link "js.fetch() (workaroun for IDEA)"
that only works in IDEA and renders as text in ASDoc HTML, and
one link "js.fetch()" that renders and works correctly in ASDoc.

aholtkamp committed:
more polyfills and actually make it work

added polyfills for:
- FormData
- Headers
- Object.assign

Only call aliasKeywordMembers after the polyfill is active

make sure that jangaroo-runtime does not use array functions or the spread operator

aholtkamp committed:
also add type definitions for Object.assign

aholtkamp committed:
added Date#toISOString()

aholtkamp committed:
added polyfill for Object.values and moved Object polyfill to jangaroo-runtime

aholtkamp committed:
fixed fetch API

aholtkamp committed:
some more API improvements

aholtkamp committed:
added polyfill for Array#from

aholtkamp commented:
Maybe we should move some of the typings to jangaroo-runtime.

aholtkamp committed:
added types and Polyfill for WeakMap

primarily to avoid loads of warning when running unit tests using phantomjs

aholtkamp committed:
added type and polyfill for globalThis

aholtkamp committed:
do not initialize globalThis

aholtkamp committed:
add typings and polyfill for URLSearchParams

fetch polyfill needs to be loaded later otherwise it will not properly detect the support for URLSearchParams and FormData

aholtkamp committed:
@see needs fully qualified name

aholtkamp committed:
switched to globalThis and added some typings

aholtkamp committed:
load Header polyfill after fetch polyfill

fwienber committed:
[jangaroo-browser] Specify [Native] name of new built-in JS types

These must be [Native] to prevent code generation in JS.
For JS, the name does not matter, but for TS conversion, the
type must called exactly like in the TS lib dom. Without the
[Native("<name>")] Annotation, the TS types would be in
namespace 'js'.
